ocaml/xapi/xapi_pool_helpers.ml

```diff
@@ -0,0 +1,76 @@
+(*
+ * Copyright (C) 2006-2009 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D=Debug.Make(struct let name="xapi" end)
+open D
+
+open Db_filter
+open Record_util
+open Threadext
+open Api_errors
+
+let all_operations = [ `ha_enable; `ha_disable ]
+
+(** Returns a table of operations -> API error options (None if the operation would be ok) *)
+let valid_operations ~__context record _ref' =
+    let _ref = Ref.string_of _ref' in
+    let current_ops = List.map snd record.Db_actions.pool_current_operations in
+    
+    let table = Hashtbl.create 10 in
+    List.iter (fun x -> Hashtbl.replace table x None) all_operations;
+    let set_errors (code: string) (params: string list) (ops: API.pool_allowed_operations_set) =
+        List.iter (fun op ->
+            if Hashtbl.find table op = None
+                then Hashtbl.replace table op (Some(code, params))) ops in
+
+    (* HA enable and disable cannot run concurrently *)
+    if List.mem `ha_enable current_ops
+        then set_errors Api_errors.ha_enable_in_progress
+            [ "pool"; _ref; pool_operation_to_string `ha_enable ] [ `ha_disable ];
+    if List.mem `ha_disable current_ops
+        then set_errors Api_errors.ha_disable_in_progress
+            [ "pool"; _ref; pool_operation_to_string `ha_disable ] [ `ha_enable ];
+    table
```

Comment:
This code looks like it will still allow multiple ha_enables or multiple ha_disables to run concurrently. We should block both operations if either one is running.